packages/terra-functional-testing/lib/services/wdio-docker-service.js

```diff
@@ -0,0 +1,167 @@
+/* eslint-disable class-methods-use-this */
+const path = require('path');
+const util = require('util');
+const childProcess = require('child_process');
+const Logger = require('../logger/logger');
+
+const exec = util.promisify(childProcess.exec);
+const logger = new Logger({ prefix: 'wdio-docker-service' });
+
+const RETRY_COUNT = 30;
+const POLL_INTERVAL = 2000;
+
+class DockerService {
+  /**
+   * Prepares the docker testing environment.
+   */
+  async onPrepare(config) {
+    this.host = config.hostname;
+    this.port = config.port;
+
+    await this.initializeSwarm();
+    await this.deployStack();
+  }
+
+  /**
+   * Initializes a docker swarm instance.
+   */
+  async initializeSwarm() {
+    const { stdout: dockerInfo } = await exec('docker info --format "{{json .}}"');
+    const { Swarm } = JSON.parse(dockerInfo);
+
+    if (Swarm.LocalNodeState !== 'active') {
+      logger.log('Initializing docker swarm.');
+
+      await exec('docker swarm init');
+    }
+  }
+
+  /**
+   * Deploys the docker stack.
+   */
+  async deployStack() {
+    // Remove the previous stack if one exists.
+    await this.removeStack();
+
+    logger.log('Deploying docker stack.');
+
+    const composeFilePath = path.resolve(__dirname, '../docker/docker-compose.yml');
+
+    await exec(`docker stack deploy -c ${composeFilePath} wdio`);
+
+    await this.awaitNetworkReady();
+  }
+
+  /**
+   * Removes the docker stack.
+   */
+  async removeStack() {
+    const { stdout: stackInfo } = await exec('docker stack ls | grep wdio || true');
```

Comment:
why `|| true` at the end of this command? If this is returned, will it attempt to remove the wdio stack when it isn't needed?

Reply:
The bash command  (`docker stack ls | grep wdio`) returns a non-zero exit code if the stack has successfully been removed because there is no output that matches the grep request. This allows the command to succeed if the stack has been removed. The alternative to this is a try catch block and analyzing the error in the catch block. A little awkward because a catch is a succeed in this case